go/multigateway/discovery_test.go

```diff
@@ -0,0 +1,851 @@
+	ctx := context.Background()
+	store := newMockStore("test-cell")
+	logger := slog.Default()
+
+	pd := NewPoolerDiscovery(ctx, store, "test-cell", logger)
+
+	// Set up initial poolers
+	pooler1 := createTestPooler("pooler1", "test-cell", "host1", "db1", "shard1", clustermetadatapb.PoolerType_PRIMARY)
+	store.conn.setInitialPoolers([]*topo.WatchDataRecursive{createPoolerWatchData(pooler1)})
+
+	// Start discovery
+	pd.Start()
+
+	// Wait for initial discovery to complete
+	waitForPoolerCount(t, pd, 1)
+
+	// Send a change
+	pooler2 := createTestPooler("pooler2", "test-cell", "host2", "db2", "shard2", clustermetadatapb.PoolerType_REPLICA)
+	store.conn.sendChange(createPoolerWatchData(pooler2))
+
+	// Wait for change to be processed
+	waitForPoolerCount(t, pd, 2)
+
+	// Stop discovery
+	pd.Stop()
+
+	// Verify discovery stopped cleanly
+	assert.Equal(t, 2, pd.PoolerCount())
+}
+
+func TestPoolerDiscovery_WatchChannelClosed(t *testing.T) {
+	t.Skip("Skipping known bug: pooler discovery doesn't reconnect after a network error")
+
+	ctx := context.Background()
+	store := newMockStore("test-cell")
+	logger := slog.Default()
+
+	pd := NewPoolerDiscovery(ctx, store, "test-cell", logger)
+
+	// Set up initial poolers
+	pooler1 := createTestPooler("pooler1", "test-cell", "host1", "db1", "shard1", clustermetadatapb.PoolerType_PRIMARY)
+	store.conn.setInitialPoolers([]*topo.WatchDataRecursive{createPoolerWatchData(pooler1)})
+
+	// Start discovery
+	pd.Start()
+
+	// Wait for initial discovery
+	waitForPoolerCount(t, pd, 1)
+
+	assert.Equal(t, 1, pd.PoolerCount(), "Should have initial pooler")
+	initialPoolers := pd.GetPoolers()
+	assert.Equal(t, "host1", initialPoolers[0].Hostname)
+
+	// Close the watch channel to simulate connection loss
+	store.conn.closeWatch()
+	store.conn.setShouldFailWatch(true)
+
+	// Wait a bit for the watch to notice the closure
+	time.Sleep(100 * time.Millisecond)
+
+	// Old values should still be cached
+	assert.Equal(t, 1, pd.PoolerCount(), "Should still have cached pooler after connection loss")
+	cachedPoolers := pd.GetPoolers()
+	assert.Equal(t, "host1", cachedPoolers[0].Hostname, "Cached values should remain")
+
+	// Now set up new poolers for after reconnection (update existing + add new)
+	pooler1Updated := createTestPooler("pooler1", "test-cell", "host1-reconnected", "db1", "shard1", clustermetadatapb.PoolerType_PRIMARY)
+	pooler2 := createTestPooler("pooler2", "test-cell", "host2", "db2", "shard2", clustermetadatapb.PoolerType_REPLICA)
+
+	// Create a new mock connection with the updated poolers
+	newConn := newMockConn("test-cell")
+	newConn.setInitialPoolers([]*topo.WatchDataRecursive{
+		createPoolerWatchData(pooler1Updated),
+		createPoolerWatchData(pooler2),
+	})
+
+	// Replace the connection so the next reconnection attempt gets new data
+	store.mu.Lock()
+	store.conn = newConn
+	store.mu.Unlock()
+
+	// Wait for reconnection and new data to be picked up
+	waitForPoolerCount(t, pd, 2)
```

Comment:
The existing bug causes this test case to fail here, since it doesn't reconnect if after a connection is lost and discover the new updates